ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+# Ruby - Phase 2
+
+This proposal is for implementing workflows in the Ruby SDK.
+
+See [phase 1](sdk-phase-1.md) for details on phase 1 which includes client and activities (granted some things may have
+changed since first authored).
+
+Notes for this proposal (like phase 1):
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+Here is a high-level overview of potentially controversial decisions:
+
+* `Temporalio::Workflow` is both the class to extend and the class with class methods for calling statically to do
+  things.
+  * The `workflow_` class methods on the `Temporalio::Workflow` can only be called at class definition time and cannot
+    be called on `Temporalio::Workflow` directly. These are for adjusting the definition.
```

Comment:
When a user writing client code writes something like

```ruby
supported_languages = wf_handle.query(
  GreetingWorkflow.<TAB>
```

will their IDE show them all the "framework" methods (including the unavailable methods starting `workflow_`). It is nice that in other languages the user's IDE only offers the methods from their own workflow interface. I think we would like to have the same be true for Ruby if possible.

Reply:
No, I am afraid this is only available at runtime (via `method_missing`). Ruby does not allow you to access an instance method at the class level.

> I think we would like to have the same be true for Ruby if possible.

Completely agree, but I cannot think of a way to do this without some form of proxying which were trying to avoid for a few reasons (mostly because we'd have to lie about what the type actually is). And even if we did do proxying, Ruby is often not expressive enough to represent this well statically. The only way I could kinda see it is if we:

```
class Workflow
  # @return [self]
  def self.stub
    StubDelegator.new(self.class)
  end
end

# This if you want to keep accurate with typing
my_wf = MyWorkflow.stub
my_wf.update('my arg') # How to use handle and args

# This if you want better options
my_wf_handle.update(MyWorkflow.stub.update, 'my arg') # So now 'update' is not the right type, checkers complain
```

Open to suggestions here, but I fear Ruby just cannot statically express things to the IDE like this. Ruby developers are used to having to use symbols or runtime methods that are not in their IDE. Of course a user can provide YARD macros or Sorbet/RBI types to support this.

Reply:
You've probably already implicitly explained this, but in Python this is done by implementing static things like `@query` decorators as functions on a `workflow` module. Can you (re-)summarize why an approach like that doesn't work for Ruby?

Reply:
In Python you can reference an _instance_ method via `Class.instance_method`, but in Ruby you cannot. You must have an instance of a class to reference an instance method on it. `FooWorkflow.my_query_method` will give a `NoMethodError`. You can do hacky things like `FooWorkflow.instance_method(:my_query_method)` to get an unbounde method, but you're into reflection there not IDE help.

Reply:
We have confirmed that even for Sorbet users, custom DSL compilers don't run in IDEs by default anyways. I cannot see a way at this time to support typesafe workflow messages without a fake/proxy type of object which I am trying to avoid. Ruby users don't really expect DSLs to be well typed in their IDE.